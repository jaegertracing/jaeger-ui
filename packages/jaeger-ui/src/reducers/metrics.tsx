// Copyright (c) 2021 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* eslint-disable camelcase */
/* eslint-disable default-case */

import { handleActions } from 'redux-actions';

import { fetchAllServiceMetrics, fetchAggregatedServiceMetrics } from '../actions/jaeger-api';
import {
  MetricPointObject,
  MetricsReduxState,
  ServiceMetrics,
  MetricObject,
  ServiceMetricsObject,
  ServiceOpsMetrics,
  OpsDataPoints,
  FetchedAllServiceMetricsResponse,
  PromiseRejectedResult,
  FetchAggregatedServiceMetricsResponse,
} from '../types/metrics';

const initialState: MetricsReduxState = {
  serviceError: {
    service_latencies_50: null,
    service_latencies_75: null,
    service_latencies_95: null,
    service_call_rate: null,
    service_error_rate: null,
  },
  opsError: {
    opsLatencies: null,
    opsCalls: null,
    opsErrors: null,
  },
  isATMActivated: null,
  loading: false,
  operationMetricsLoading: null,
  serviceMetrics: null,
  serviceOpsMetrics: undefined,
};

function fetchStarted(state: MetricsReduxState) {
  return {
    ...state,
    serviceError: initialState.serviceError,
    loading: true,
  };
}

function fetchServiceMetricsDone(
  state: MetricsReduxState,
  { payload }: { payload?: FetchedAllServiceMetricsResponse }
) {
  const serviceMetrics: ServiceMetrics = {
    service_latencies: null,
    service_call_rate: null,
    service_error_rate: null,
  };
  const serviceError: MetricsReduxState['serviceError'] = {
    service_latencies_50: null,
    service_latencies_75: null,
    service_latencies_95: null,
    service_call_rate: null,
    service_error_rate: null,
  };
  let isATMActivated = true;

  if (payload) {
    payload.forEach((promiseResult, i) => {
      if (promiseResult.status === 'fulfilled') {
        const metrics = promiseResult.value;
        if (metrics.metrics[0]) {
          let max = 0;
          const metric: ServiceMetricsObject = {
            serviceName: metrics.metrics[0].labels[0].value,
            quantile: metrics.quantile,
            max: 0,
            metricPoints: metrics.metrics[0].metricPoints.map((p: MetricPointObject) => {
              let y;
              try {
                y = parseFloat(p.gaugeValue.doubleValue.toFixed(2));
                max = y > max ? y : max;
              } catch (e) {
                y = null;
              }
              return {
                x: new Date(p.timestamp).getTime(),
                y,
              };
            }),
          };

          metric.max = max;
          if (metrics.name === 'service_latencies') {
            if (serviceMetrics[metrics.name] === null) {
              serviceMetrics[metrics.name] = [];
            }
            serviceMetrics[metrics.name]!.push(metric);
          } else {
            serviceMetrics[metrics.name] = metric;
          }
        }
      } else {
        if (
          typeof (promiseResult as PromiseRejectedResult).reason === 'object' &&
          (promiseResult as PromiseRejectedResult).reason.httpStatus === 501
        ) {
          isATMActivated = false;
        }

        switch (i) {
          case 0:
            serviceError.service_latencies_50 = (promiseResult as PromiseRejectedResult).reason;
            break;
          case 1:
            serviceError.service_latencies_75 = (promiseResult as PromiseRejectedResult).reason;
            break;
          case 2:
            serviceError.service_latencies_95 = (promiseResult as PromiseRejectedResult).reason;
            break;
          case 3:
            serviceError.service_call_rate = (promiseResult as PromiseRejectedResult).reason;
            break;
          case 4:
            serviceError.service_error_rate = (promiseResult as PromiseRejectedResult).reason;
            break;
        }
      }
    });
  }
  return { ...state, serviceMetrics, serviceError, loading: false, isATMActivated };
}

function fetchOpsMetricsStarted(state: MetricsReduxState) {
  return {
    ...state,
    opsError: initialState.opsError,
    operationMetricsLoading: true,
  };
}

function fetchOpsMetricsDone(
  state: MetricsReduxState,
  { payload }: { payload?: FetchAggregatedServiceMetricsResponse }
) {
  const opsError: MetricsReduxState['opsError'] = {
    opsLatencies: null,
    opsCalls: null,
    opsErrors: null,
  };

  let opsMetrics:
    | null
    | {
        [k in string]: {
          name: string;
          metricPoints: OpsDataPoints;
        };
      } = null;

  // eslint-disable-next-line no-undef-init
  let serviceOpsMetrics: ServiceOpsMetrics[] | undefined = undefined;

  if (payload) {
    payload.forEach((promiseResult, i) => {
      if (promiseResult.status === 'fulfilled') {
        const metric = promiseResult.value;
        metric.metrics.forEach((metricDetails: MetricObject) => {
          if (opsMetrics === null) {
            opsMetrics = {};
          }

          let opsName: string | null = null;
          const avg: {
            service_operation_latencies: number;
            service_operation_call_rate: number;
            service_operation_error_rate: number;
          } = {
            service_operation_latencies: 0,
            service_operation_call_rate: 0,
            service_operation_error_rate: 0,
          };
          metricDetails.labels.forEach((label: any) => {
            if (label.name === 'operation') {
              opsName = label.value;
            }
          });

          if (opsName) {
            if (opsMetrics[opsName] === undefined) {
              opsMetrics[opsName] = {
                name: opsName,
                metricPoints: {
                  service_operation_latencies: [],
                  service_operation_call_rate: [],
                  service_operation_error_rate: [],
                  avg: {
                    service_operation_latencies: null,
                    service_operation_call_rate: null,
                    service_operation_error_rate: null,
                  },
                },
              };
            }

            opsMetrics[opsName].metricPoints[metric.name] = metricDetails.metricPoints.map(p => {
              let y;
              try {
                y = parseFloat(p.gaugeValue.doubleValue.toFixed(2));
                avg[metric.name] += y;
              } catch (e) {
                y = null;
              }

              return {
                x: new Date(p.timestamp).getTime(),
                y,
              };
            });

            opsMetrics[opsName].metricPoints.avg[metric.name] =
              metricDetails.metricPoints.length > 0
                ? parseFloat((avg[metric.name] / metricDetails.metricPoints.length).toFixed(2))
                : null;
          }
        });
      } else {
        switch (i) {
          case 0:
            opsError.opsLatencies = (promiseResult as PromiseRejectedResult).reason;
            break;
          case 1:
            opsError.opsCalls = (promiseResult as PromiseRejectedResult).reason;
            break;
          case 2:
            opsError.opsErrors = (promiseResult as PromiseRejectedResult).reason;
            break;
        }
      }
    });

    const minMax: {
      min: number;
      max: number;
    } = {
      min: 0,
      max: 0,
    };

    if (opsMetrics) {
      serviceOpsMetrics = Object.keys(opsMetrics).map((operationName, i) => {
        let impact = 0;
        if (
          opsMetrics![operationName].metricPoints.avg.service_operation_latencies !== null &&
          opsMetrics![operationName].metricPoints.avg.service_operation_call_rate !== null
        ) {
          impact =
            (opsMetrics![operationName].metricPoints.avg.service_operation_latencies! *
              opsMetrics![operationName].metricPoints.avg.service_operation_call_rate!) /
            100;
        }

        if (i === 0) {
          minMax.max = impact;
          minMax.min = impact;
        } else {
          minMax.max = minMax.max < impact ? impact : minMax.max;
          minMax.min = minMax.min > impact ? impact : minMax.min;
        }

        return {
          key: i,
          name: opsMetrics![operationName].name,
          latency: opsMetrics![operationName].metricPoints.avg.service_operation_latencies || 0,
          requests: opsMetrics![operationName].metricPoints.avg.service_operation_call_rate || 0,
          errRates: opsMetrics![operationName].metricPoints.avg.service_operation_error_rate || 0,
          impact,
          dataPoints: opsMetrics![operationName].metricPoints,
        };
      });

      if (serviceOpsMetrics && serviceOpsMetrics.length === 1) {
        serviceOpsMetrics.forEach((v, i) => {
          serviceOpsMetrics![i].impact = 1;
        });
      } else if (serviceOpsMetrics && serviceOpsMetrics.length > 1) {
        serviceOpsMetrics.forEach((v, i) => {
          if (minMax.max - minMax.min === 0) {
            serviceOpsMetrics![i].impact = 0;
          } else {
            serviceOpsMetrics![i].impact = (v.impact - minMax.min) / (minMax.max - minMax.min);
          }
        });
      }
    }
  }

  return { ...state, serviceOpsMetrics, opsError, operationMetricsLoading: false };
}

export default handleActions(
  {
    [`${fetchAllServiceMetrics}_PENDING`]: fetchStarted,
    [`${fetchAllServiceMetrics}_FULFILLED`]: fetchServiceMetricsDone,

    [`${fetchAggregatedServiceMetrics}_PENDING`]: fetchOpsMetricsStarted,
    [`${fetchAggregatedServiceMetrics}_FULFILLED`]: fetchOpsMetricsDone,
  },
  initialState
);
