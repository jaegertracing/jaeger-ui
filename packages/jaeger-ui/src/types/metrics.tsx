// Copyright (c) 2021 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* eslint-disable camelcase */

import { ApiError } from './api-error';

export type MetricsType = 'latencies' | 'calls' | 'errors';
export type AvailableServiceMetrics = 'service_call_rate' | 'service_latencies' | 'service_error_rate';
export type AvailableOpsMetrics =
  | 'service_operation_call_rate'
  | 'service_operation_latencies'
  | 'service_operation_error_rate';

export type MetricsAPIQueryParams = {
  quantile: number;
  groupByOperation?: boolean;
  endTs?: number;
  lookback?: number;
  step?: number;
  ratePer?: number;
  spanKind?: 'unspecified' | 'internal' | 'server' | 'client' | 'producer' | 'consumer';
};

export type LableObject = {
  name: string;
  value: string;
};

export type MetricPointObject = {
  gaugeValue: {
    doubleValue: number;
  };
  timestamp: string;
};

export type MetricObject = {
  labels: LableObject[];
  metricPoints: MetricPointObject[];
};

export type MetricsAPIServiceResponseData<T = AvailableServiceMetrics, U = 0.95> = {
  name: T;
  type: 'GAUGE';
  help: string;
  metrics: MetricObject[];
  quantile: U;
};

export type MetricsAPIOpsResponseData<T = AvailableOpsMetrics> = {
  name: T;
  type: 'GAUGE';
  help: string;
  metrics: MetricObject[];
  quantile: number;
};

export type Points = {
  x: number;
  y: number | null;
};

export type DataAvg = {
  service_operation_call_rate: null | number;
  service_operation_error_rate: null | number;
  service_operation_latencies: null | number;
};

export type OpsDataPoints = {
  service_operation_call_rate: Points[];
  service_operation_error_rate: Points[];
  service_operation_latencies: Points[];
  avg: DataAvg;
};

export type ServiceOpsMetrics = {
  dataPoints: OpsDataPoints;
  errRates: number;
  impact: number;
  latency: number;
  name: string;
  requests: number;
  key: number;
};

export type ServiceMetricsObject = {
  serviceName: string;
  quantile: number;
  max: number;
  metricPoints: Points[];
};

export type ServiceMetrics = {
  service_latencies: null | ServiceMetricsObject[];
  service_call_rate: null | ServiceMetricsObject;
  service_error_rate: null | ServiceMetricsObject;
};

export type MetricsReduxState = {
  serviceError: {
    service_latencies_50: null | ApiError;
    service_latencies_75: null | ApiError;
    service_latencies_95: null | ApiError;
    service_call_rate: null | ApiError;
    service_error_rate: null | ApiError;
  };
  opsError: {
    opsLatencies: null | ApiError;
    opsCalls: null | ApiError;
    opsErrors: null | ApiError;
  };
  isATMActivated: null | boolean;
  loading: boolean;
  operationMetricsLoading: null | boolean;
  serviceMetrics: ServiceMetrics | null;
  serviceOpsMetrics: ServiceOpsMetrics[] | undefined;
};

export enum PromiseStatus {
  fulfilled = 'fulfilled',
  rejected = 'rejected',
}

export type PromiseFulfilledResult<T> = {
  status: PromiseStatus.fulfilled;
  value: T;
};

export type PromiseRejectedResult = {
  status: PromiseStatus.rejected;
  reason: any;
};

export type FetchedAllServiceMetricsResponse = [
  PromiseFulfilledResult<MetricsAPIServiceResponseData<'service_latencies', 0.5>> | PromiseRejectedResult,
  PromiseFulfilledResult<MetricsAPIServiceResponseData<'service_latencies', 0.75>> | PromiseRejectedResult,
  PromiseFulfilledResult<MetricsAPIServiceResponseData<'service_latencies'>> | PromiseRejectedResult,
  PromiseFulfilledResult<MetricsAPIServiceResponseData<'service_call_rate'>> | PromiseRejectedResult,
  PromiseFulfilledResult<MetricsAPIServiceResponseData<'service_error_rate'>> | PromiseRejectedResult
];

export type FetchAggregatedServiceMetricsResponse = [
  PromiseFulfilledResult<MetricsAPIOpsResponseData<'service_operation_latencies'>> | PromiseRejectedResult,
  PromiseFulfilledResult<MetricsAPIOpsResponseData<'service_operation_call_rate'>> | PromiseRejectedResult,
  PromiseFulfilledResult<MetricsAPIOpsResponseData<'service_operation_error_rate'>> | PromiseRejectedResult
];
