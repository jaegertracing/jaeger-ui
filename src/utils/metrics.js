// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import ReactGA from 'react-ga';

export function init() {
  if (process.env.NODE_ENV === 'production' && process.env.REACT_APP_GA_TRACKING_ID) {
    const GA_CODE = process.env.REACT_APP_GA_TRACKING_ID;
    ReactGA.initialize(GA_CODE);
  }
}

export function trackPageView(pathname, search) {
  const pagePath = search ? `${pathname}?${search}` : pathname;
  ReactGA.pageview(pagePath);
}
