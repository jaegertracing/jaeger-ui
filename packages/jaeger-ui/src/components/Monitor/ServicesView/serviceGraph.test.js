// Copyright (c) 2021 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import ServiceGraph, { tickFormat } from './serviceGraph';
import { serviceMetrics } from '../../../reducers/metrics.mock';

const props = {
  width: 300,
  error: null,
  name: 'Hello Graph',
  metricsData: null,
  loading: true,
  marginClassName: '',
  crosshairValues: [],
  xDomain: [1, 2],
};

describe('<ServiceGraph>', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ServiceGraph {...props} />);
  });

  it('does not explode', () => {
    expect(wrapper.length).toBe(1);
  });

  it('Loading indicator is displayed', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Loading indicator is displayed when xDomain is empty', () => {
    wrapper.setProps({ ...props, xDomain: [], loading: false });
    expect(wrapper).toMatchSnapshot();
  });

  it('"No data" displayed', () => {
    wrapper.setProps({ ...props, loading: false });
    expect(wrapper).toMatchSnapshot();
  });

  it('"Couldnâ€™t fetch data" displayed', () => {
    wrapper.setProps({ ...props, loading: false, error: new Error('API Error') });
    expect(wrapper).toMatchSnapshot();
  });

  it('Base graph should be displayed ', () => {
    wrapper.setProps({ ...props, loading: false, metricsData: serviceMetrics.service_call_rate });
    expect(wrapper).toMatchSnapshot();
  });

  it('Base graph with legends should be displayed', () => {
    wrapper.setProps({
      ...props,
      loading: false,
      metricsData: serviceMetrics.service_call_rate,
      showLegend: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Base graph with horizontal lines should be displayed', () => {
    wrapper.setProps({
      ...props,
      loading: false,
      metricsData: serviceMetrics.service_call_rate,
      showHorizontalLines: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Base graph with custom color should be displayed', () => {
    wrapper.setProps({
      ...props,
      loading: false,
      metricsData: serviceMetrics.service_call_rate,
      color: 'AAAAAA',
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Crosshair map test', () => {
    wrapper.setProps({
      ...props,
      loading: false,
      metricsData: serviceMetrics.service_call_rate,
      crosshairValues: [
        {
          key: 1,
          ...serviceMetrics.service_call_rate.metricPoints[0],
        },
      ],
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('Crosshair mouse hover test', () => {
    wrapper.setProps({
      ...props,
      loading: false,
      metricsData: {
        ...serviceMetrics.service_call_rate,
        metricPoints: [{ x: 1631271783806, y: null }, ...serviceMetrics.service_call_rate.metricPoints],
      },
      crosshairValues: [
        {
          key: 1,
          ...serviceMetrics.service_call_rate.metricPoints[0],
        },
      ],
    });
    expect(wrapper.find('AreaSeries').render()).toMatchSnapshot();
    expect(wrapper.find('LineSeries').render()).toMatchSnapshot();
  });

  it('AreaSeries onNearestX, XYPlot onMouseLeave', () => {
    wrapper.setProps({
      ...props,
      loading: false,
      metricsData: serviceMetrics.service_call_rate,
    });
    wrapper
      .find('AreaSeries')
      .at(0)
      .prop('onNearestX')({ x: 1, y: 2 }, { index: 7 });
    expect(wrapper.state().crosshairValues).toEqual([{ label: 0.95 }]);
    wrapper.find('XYPlot').prop('onMouseLeave')();
    expect(wrapper.state().crosshairValues).toEqual([]);
  });
});

describe('<ServiceGraph> util', () => {
  it('tickFormat singular', () => {
    expect(tickFormat(Date.UTC(2017, 1, 14, 15, 19)).valueOf()).toBe('15:19');
  });

  it('tickFormat plural', () => {
    expect(tickFormat(Date.UTC(2017, 1, 14, 4, 4)).valueOf()).toBe('04:04');
  });
});
