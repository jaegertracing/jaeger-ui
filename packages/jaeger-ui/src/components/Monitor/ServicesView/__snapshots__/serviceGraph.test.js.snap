// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ServiceGraph> "Couldn’t fetch data" displayed 1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <div
    className="center-placeholder"
    style={
      Object {
        "height": 168,
        "width": 300,
      }
    }
  >
    Couldn’t fetch data
  </div>
</div>
`;

exports[`<ServiceGraph> "No data" displayed 1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <div
    className="center-placeholder"
    style={
      Object {
        "height": 168,
        "width": 300,
      }
    }
  >
    No Data
  </div>
</div>
`;

exports[`<ServiceGraph> Base graph should be displayed  1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <XYPlot
    className=""
    height={168}
    margin={
      Object {
        "bottom": 25,
      }
    }
    onMouseLeave={[Function]}
    width={300}
    xDomain={
      Array [
        1,
        2,
      ]
    }
  >
    <XAxis
      attr="x"
      attrAxis="y"
      orientation="bottom"
      tickFormat={[Function]}
      tickTotal={5}
    />
    <YAxis
      attr="y"
      attrAxis="x"
      orientation="left"
    />
    <AreaSeries
      className=""
      color={
        Array [
          "#DCA3D2",
        ]
      }
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="0"
      onNearestX={[Function]}
      opacity={0.1}
      stack={false}
      style={Object {}}
    />
    <LineSeries
      className=""
      color={
        Array [
          "#DCA3D2",
        ]
      }
      curve={null}
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="1"
      opacity={1}
      stack={false}
      strokeStyle="solid"
      style={Object {}}
    />
    <Crosshair
      itemsFormat={[Function]}
      style={
        Object {
          "box": Object {},
          "line": Object {},
          "title": Object {},
        }
      }
      titleFormat={[Function]}
      values={Array []}
    >
      <div
        className="crosshair-value"
      />
    </Crosshair>
  </XYPlot>
</div>
`;

exports[`<ServiceGraph> Base graph with custom color should be displayed 1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <XYPlot
    className=""
    height={168}
    margin={
      Object {
        "bottom": 25,
      }
    }
    onMouseLeave={[Function]}
    width={300}
    xDomain={
      Array [
        1,
        2,
      ]
    }
  >
    <XAxis
      attr="x"
      attrAxis="y"
      orientation="bottom"
      tickFormat={[Function]}
      tickTotal={5}
    />
    <YAxis
      attr="y"
      attrAxis="x"
      orientation="left"
    />
    <AreaSeries
      className=""
      color={
        Array [
          "AAAAAA",
        ]
      }
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="0"
      onNearestX={[Function]}
      opacity={0.1}
      stack={false}
      style={Object {}}
    />
    <LineSeries
      className=""
      color={
        Array [
          "AAAAAA",
        ]
      }
      curve={null}
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="1"
      opacity={1}
      stack={false}
      strokeStyle="solid"
      style={Object {}}
    />
    <Crosshair
      itemsFormat={[Function]}
      style={
        Object {
          "box": Object {},
          "line": Object {},
          "title": Object {},
        }
      }
      titleFormat={[Function]}
      values={Array []}
    >
      <div
        className="crosshair-value"
      />
    </Crosshair>
  </XYPlot>
</div>
`;

exports[`<ServiceGraph> Base graph with horizontal lines should be displayed 1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <XYPlot
    className=""
    height={168}
    margin={
      Object {
        "bottom": 25,
      }
    }
    onMouseLeave={[Function]}
    width={300}
    xDomain={
      Array [
        1,
        2,
      ]
    }
  >
    <HorizontalGridLines
      attr="y"
      direction="horizontal"
    />
    <XAxis
      attr="x"
      attrAxis="y"
      orientation="bottom"
      tickFormat={[Function]}
      tickTotal={5}
    />
    <YAxis
      attr="y"
      attrAxis="x"
      orientation="left"
    />
    <AreaSeries
      className=""
      color={
        Array [
          "#DCA3D2",
        ]
      }
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="0"
      onNearestX={[Function]}
      opacity={0.1}
      stack={false}
      style={Object {}}
    />
    <LineSeries
      className=""
      color={
        Array [
          "#DCA3D2",
        ]
      }
      curve={null}
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="1"
      opacity={1}
      stack={false}
      strokeStyle="solid"
      style={Object {}}
    />
    <Crosshair
      itemsFormat={[Function]}
      style={
        Object {
          "box": Object {},
          "line": Object {},
          "title": Object {},
        }
      }
      titleFormat={[Function]}
      values={Array []}
    >
      <div
        className="crosshair-value"
      />
    </Crosshair>
  </XYPlot>
</div>
`;

exports[`<ServiceGraph> Base graph with legends should be displayed 1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <XYPlot
    className=""
    height={168}
    margin={
      Object {
        "bottom": 25,
      }
    }
    onMouseLeave={[Function]}
    width={300}
    xDomain={
      Array [
        1,
        2,
      ]
    }
  >
    <XAxis
      attr="x"
      attrAxis="y"
      orientation="bottom"
      tickFormat={[Function]}
      tickTotal={5}
    />
    <YAxis
      attr="y"
      attrAxis="x"
      orientation="left"
    />
    <AreaSeries
      className=""
      color={
        Array [
          "#DCA3D2",
        ]
      }
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="0"
      onNearestX={[Function]}
      opacity={0.1}
      stack={false}
      style={Object {}}
    />
    <LineSeries
      className=""
      color={
        Array [
          "#DCA3D2",
        ]
      }
      curve={null}
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="1"
      opacity={1}
      stack={false}
      strokeStyle="solid"
      style={Object {}}
    />
    <Crosshair
      itemsFormat={[Function]}
      style={
        Object {
          "box": Object {},
          "line": Object {},
          "title": Object {},
        }
      }
      titleFormat={[Function]}
      values={Array []}
    >
      <div
        className="crosshair-value"
      />
    </Crosshair>
    <DiscreteColorLegendItem
      className="legend-label"
      colors={
        Array [
          "#12939A",
          "#79C7E3",
          "#1A3177",
          "#FF9833",
          "#EF5D28",
        ]
      }
      items={
        Array [
          Object {
            "color": "#DCA3D2",
            "title": "95th",
          },
        ]
      }
      orientation="horizontal"
    />
  </XYPlot>
</div>
`;

exports[`<ServiceGraph> Crosshair map test 1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <XYPlot
    className=""
    height={168}
    margin={
      Object {
        "bottom": 25,
      }
    }
    onMouseLeave={[Function]}
    width={300}
    xDomain={
      Array [
        1,
        2,
      ]
    }
  >
    <XAxis
      attr="x"
      attrAxis="y"
      orientation="bottom"
      tickFormat={[Function]}
      tickTotal={5}
    />
    <YAxis
      attr="y"
      attrAxis="x"
      orientation="left"
    />
    <AreaSeries
      className=""
      color={
        Array [
          "#DCA3D2",
        ]
      }
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="0"
      onNearestX={[Function]}
      opacity={0.1}
      stack={false}
      style={Object {}}
    />
    <LineSeries
      className=""
      color={
        Array [
          "#DCA3D2",
        ]
      }
      curve={null}
      data={
        Array [
          Object {
            "x": 1631271823806,
            "y": 0.05,
          },
          Object {
            "x": 1631271883806,
            "y": 0.05,
          },
        ]
      }
      getNull={[Function]}
      key="1"
      opacity={1}
      stack={false}
      strokeStyle="solid"
      style={Object {}}
    />
    <Crosshair
      itemsFormat={[Function]}
      style={
        Object {
          "box": Object {},
          "line": Object {},
          "title": Object {},
        }
      }
      titleFormat={[Function]}
      values={Array []}
    >
      <div
        className="crosshair-value"
      />
    </Crosshair>
  </XYPlot>
</div>
`;

exports[`<ServiceGraph> Crosshair mouse hover test 1`] = `
<path
  class="rv-xy-plot__series rv-xy-plot__series--line "
  d="M0,0L0,0L0,0L0,0Z"
  style="opacity:0.1;stroke:#DCA3D2;fill:#DCA3D2"
  transform="translate(undefined,undefined)"
/>
`;

exports[`<ServiceGraph> Crosshair mouse hover test 2`] = `
<path
  class="rv-xy-plot__series rv-xy-plot__series--line "
  d="M0,0L0,0"
  style="opacity:1;stroke:#DCA3D2"
  transform="translate(undefined,undefined)"
/>
`;

exports[`<ServiceGraph> Loading indicator is displayed 1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <div
    className="center-placeholder"
    style={
      Object {
        "height": 168,
        "width": 300,
      }
    }
  >
    <LoadingIndicator
      centered={true}
      small={false}
    />
  </div>
</div>
`;

exports[`<ServiceGraph> Loading indicator is displayed when xDomain is empty 1`] = `
<div
  className="graph-container "
  style={
    Object {
      "height": 242,
    }
  }
>
  <h3
    className="graph-header"
  >
    Hello Graph
  </h3>
  <div
    className="center-placeholder"
    style={
      Object {
        "height": 168,
        "width": 300,
      }
    }
  >
    <LoadingIndicator
      centered={true}
      small={false}
    />
  </div>
</div>
`;
